restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +91,90 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
     }
   }
 
+  /**
+   *
+   * @param dataSchema
+   * @param dataMap
+   * @return a new different data map that contains original data plus the default values
+   */
+  public static DataMap fillInDefaultValues(DataSchema dataSchema, DataMap dataMap)
+  {
+    DataMap dataDefaultFilled = new DataMap();
+    if (dataMap != null)
+    {
+      dataDefaultFilled.putAll(dataMap);
+    }
+
+    if (dataSchema instanceof TyperefDataSchema)
+    {
+      TyperefDataSchema typerefDataSchema = (TyperefDataSchema) dataSchema;
+      return fillInDefaultValues(typerefDataSchema.getDereferencedDataSchema(), dataDefaultFilled);
+    }
+    else if (dataSchema instanceof RecordDataSchema)
+    {
+      RecordDataSchema recordDataSchema = (RecordDataSchema) dataSchema;
+      for (RecordDataSchema.Field field : recordDataSchema.getFields())
+      {
+        DataSchema fieldSchema = field.getType();
+        if (fieldSchema instanceof RecordDataSchema)
```

Comment:
Why duplicate the logic for checking Record/Typeref again? Why not just make a recursive call for each field?

Reply:
good suggestion

Reply:
I found here it is needed to differentiate the case but I do changed the code for the redundant logic of checking assigned field. but for this part about "record field and typeref field logic redundancy" I feel I can not change it because there are some logic necessary, let me know!

Reply:
I think you can simplify the logic by reversing the checks.

First check and compute the field value:
if( dataDefaultFilled.containsKey())
   Object fieldData = // get data from datamap
else if field has default
  fieldData = // get field default.
else
  // continue (nothing to do).

Then compute the default for the field by calling recursively:
if fieldData instaceOf DataMap || instance of DataList
  fieldData = fillInDefaults(fieldData, fieldSchema)

dataDefaultFilled.put(fieldName, fieldData)